src/NuGet.Clients/PackageManagement.VisualStudio/Migrate/ProjectJsonToPackageRefMigrator.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using EnvDTE;
+using Microsoft.VisualStudio.Shell;
+using NuGet.Common;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectModel;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public static class ProjectJsonToPackageRefMigrator
+    {
+        public static async Task<ProjectJsonToPackageRefMigrateResult> MigrateAsync(LegacyCSProjPackageReferenceProject project, string dteProjectFullName)
+        {
+            var projectJsonFilePath = ProjectJsonPathUtilities.GetProjectConfigPath(Path.GetDirectoryName(project.MSBuildProjectPath),
+                Path.GetFileNameWithoutExtension(project.MSBuildProjectPath));
+
+            var packageSpec = JsonPackageSpecReader.GetPackageSpec(
+                Path.GetFileNameWithoutExtension(project.MSBuildProjectPath),
+                projectJsonFilePath);
+
+            await MigrateDependencies(project, packageSpec);
+
+            var buildProject = EnvDTEProjectUtility.AsMicrosoftBuildEvaluationProject(dteProjectFullName);
```

Comment:
Current guideline is to always access this `buildProject` on UI thread otherwise it will be vulnerable. So make sure wherever you're using this msbuild project, you move to UI thread explicitly